client/client.go

```diff
@@ -1944,6 +1947,7 @@ OUTER:
 			filtered:      filtered,
 			pulled:        pulledAllocs,
 			migrateTokens: resp.MigrateTokens,
+			index:         resp.Index,
```

Comment:
L1942  updates req.MinQueryIndex if and only if resp.Index is greater, so I wonder if there's some reason we should use req.MinQueryIndex here instead. I'm honestly not sure L1942 is reachable. Perhaps there's a timeout that could cause a response before resp.Index is greater than MinQueryIndex?

Not a blocker as I think at worst it's an edge case of an edge case that when hit will negate the correctness improvement of this PR. It can't make the behavior _worse_ than before the PR AFAICT.

Reply:
I don't think we should be using `req.MinQueryIndex` here.  It's simpler to reason that we reconciling local state against pulled state (at index resp.Index`) without worrying much about the indirection or interference of `req.MinQueryIndex` (i.e. if resp.Index is earlier than req.MinQueryIndex, using req.MinQueryIndex risks us believing the server state is more recent than it actually is).  We expect the reconciler to work even if resp.Index went back in time expectedly.

As for `req.MinQueryIndex`, it seems that we are protecting against servers state going back in time!  That feels quite odd and I wonder if it's just being defensive or a case we hit at some point.